Benchmarks/benchmark.cpp

```diff
@@ -0,0 +1,281 @@
+			{
+				state.PauseTiming();
+				reader.close();
+				reader.open();
+				state.ResumeTiming();
+			}
+		}
+
+		++totalPackets;
+		totalBytes += rawPacket.getRawDataLen();
+	}
+
+	state.SetBytesProcessed(totalBytes);
+	state.SetItemsProcessed(totalPackets);
+}
+BENCHMARK(BM_PcapFileRead);
+
+static void BM_PcapFileWrite(benchmark::State& state)
+{
+	// Open the pcap file for writing
+	pcpp::PcapFileWriterDevice writer("PcapExamples/output.pcap");
+	if (!writer.open())
+	{
+		state.SkipWithError("Cannot open pcap file for writing");
+		return;
+	}
+
+	pcpp::Packet packet;
+	pcpp::EthLayer ethLayer(pcpp::MacAddress("00:00:00:00:00:00"), pcpp::MacAddress("00:00:00:00:00:00"));
+	pcpp::IPv4Layer ip4Layer(pcpp::IPv4Address("192.168.0.1"), pcpp::IPv4Address("192.168.0.2"));
+	pcpp::TcpLayer tcpLayer(12345, 80);
+
+	packet.addLayer(&ethLayer);
+	packet.addLayer(&ip4Layer);
+	packet.addLayer(&tcpLayer);
+	packet.computeCalculateFields();
+
+	size_t totalBytes = 0;
+	size_t totalPackets = 0;
+	for (auto _ : state)
+	{
+		// Write packet to file
+		writer.writePacket(*(packet.getRawPacket()));
+
+		// Count total bytes and packets
+		++totalPackets;
+		totalBytes += packet.getRawPacket()->getRawDataLen();
+	}
+
+	// Set statistics to the benchmark state
+	state.SetBytesProcessed(totalBytes);
+	state.SetItemsProcessed(totalPackets);
+}
+BENCHMARK(BM_PcapFileWrite);
+
+static void BM_PacketParsing(benchmark::State& state)
+{
+	std::unordered_map<pcpp::ProtocolType, size_t> layerTypes;
+	std::unordered_map<pcpp::OsiModelLayer, size_t> osiLayers;
+
+	// Open the pcap file for reading
+	size_t totalBytes = 0;
+	size_t totalPackets = 0;
+	pcpp::PcapFileReaderDevice reader(pcapFileName);
+	if (!reader.open())
+	{
+		state.SkipWithError("Cannot open pcap file for reading");
+		return;
+	}
+
+	pcpp::RawPacket rawPacket;
+	for (auto _ : state)
+	{
+		if (!reader.getNextPacket(rawPacket))
+		{
+			// If the rawPacket is empty there should be an error
+			if (totalBytes == 0)
+			{
+				state.SkipWithError("Cannot read packet");
+				break;
+			}
+			else  // If the rawPacket is not empty, it means the file is over
+			{
+				state.PauseTiming();
+				reader.close();
+				reader.open();
+				state.ResumeTiming();
+			}
+		}
+
+		// Parse packet
+		pcpp::Packet parsedPacket(&rawPacket);
+
+		// Count protocol layers to simulate accessing
+		for (pcpp::Layer* curLayer = parsedPacket.getFirstLayer(); curLayer != NULL;
+		     curLayer = curLayer->getNextLayer())
+		{
+			// Count protocol types. If the protocol type is not in the map, add it
+			if (layerTypes.find(curLayer->getProtocol()) == layerTypes.end())
+			{
+				layerTypes[curLayer->getProtocol()] = 0;
+			}
+			++layerTypes[curLayer->getProtocol()];
+
+			// Count OSI layers. If the OSI layer is not in the map, add it
+			if (osiLayers.find(curLayer->getOsiModelLayer()) == osiLayers.end())
+			{
+				osiLayers[curLayer->getOsiModelLayer()] = 0;
+			}
+			++osiLayers[curLayer->getOsiModelLayer()];
+		}
```

Comment:
If we want to measure the packet parsing, why do we need this? It's enough to just parse the packet, no? 🤔

Reply:
My concern is if some compiler decides the parsedPacket is not used it might try to remove the parsing maybe. But I'm not sure

Reply:
maybe we can do something simple like asserting the first layer is not null? 🤔